modules/ecs-service/main.tf

```diff
@@ -0,0 +1,696 @@
+
+locals {
+  # If the container image is not specified, then the latest version of the
+  # container definition image that is deployed will be used.
+  lookup_deployed_image = var.ecs_container_image == ""
+
+  # When looking up the deployed image, we need to distinguish between
+  # a scheduled task and a service. A scheduled task will not have a
+  # service associated with it
+  image = (
+    local.lookup_deployed_image ?
+    (
+      var.create_scheduled_task ?
+      # data.aws_ecs_container_definition.scheduled[0].image :
+      var.ecs_container_image : # TODO FIX THIS
+      data.aws_ecs_container_definition.service[0].image
+    )
+    : var.ecs_container_image
+  )
+}
+
+# -------------------------------------------
+# LOOKUP IMAGE FOR ECS SERVICE
+# -------------------------------------------
+
+# This will get the ECS service that is currently deployed.
+data "aws_ecs_service" "service" {
+  count = !var.create_scheduled_task && local.lookup_deployed_image ? 1 : 0
+
+  cluster_arn  = data.aws_ecs_cluster.cluster.arn
+  service_name = var.ecs_service_name
+}
+
+# This will get the latest container definition from the task definition
+# that is currently deployed to the ECS service.
+data "aws_ecs_container_definition" "service" {
+  count = !var.create_scheduled_task && local.lookup_deployed_image ? 1 : 0
+
+  task_definition = data.aws_ecs_service.service[0].task_definition
+  container_name  = var.ecs_service_name
+}
+
+
+# -------------------------------------------
+# LOOKUP IMAGE FOR SCHEDULED TASK
+# -------------------------------------------
+
+# TODO: This causes a circular dependency, new solution is needed
+
+# Gets the latest container definition from the max
+# revision of the task definition.
+# data "aws_ecs_container_definition" "scheduled" {
+#   count = var.create_scheduled_task && local.lookup_deployed_image ? 1 : 0
+
+#   task_definition = local.task_definition
+#   container_name  = var.ecs_service_name
+
+#   depends_on = [
+#     data.aws_ecs_task_definition.task
+#   ]
+# }
+
+
+# -------------------------------------------
+# CREATE THE ECS TASK DEFINITION
+# -------------------------------------------
+
+locals {
+  log_group_name = "/ecs/service/${var.ecs_service_name}"
+  log_configuration = {
+    logDriver = "awslogs"
+    options = {
+      awslogs-group         = local.log_group_name
+      awslogs-region        = data.aws_region.current.name
+      awslogs-stream-prefix = "stream"
+    }
+  }
+
+  repositoryCredentials = {
+    credentialsParameter = var.docker_credential_secretsmanager_arn
+  }
+}
+
+resource "aws_ecs_task_definition" "task" {
+  family = var.ecs_service_name
+
+  cpu          = 256
+  memory       = 256
+  network_mode = var.create_scheduled_task ? "awsvpc" : "bridge"
+
+  execution_role_arn = aws_iam_role.task_execution.arn
+  task_role_arn      = length(var.ecs_task_role_policy_arns) > 0 ? aws_iam_role.task[0].arn : null
+
+  container_definitions = jsonencode([
```

Comment:
data container definition only allows us to query existing container definitions not define them like how IAM uses policy data blocks. This is a discrepancy with the aws provider, some services follow the broken out paradigm except for ecs... This is why we are doing the bootstrapping, otherwise we wouldn't have a problem to work around, this is a long running complaint with ecs service part of the provider.